### PR TITLE
Implement new captcha configuration

### PR DESCRIPTION
**Pull request by pkong-ds**
ref #4325 

@louischan-oursky thanks, copied items from issue below

>  Keep the old configuration intact
>  Introduce the new configuration
>>        The Go structs
>>        The JSON schema
>>        Write tests to test for valid and invalid configurations

### CODE REVIEW

File: pkg/lib/config/captcha.go [hunk 1]
```diff
@@ -5,24 +5,59 @@ var _ = Schema.Add("CaptchaConfig", `
...
+
+var _ = Schema.Add("CaptchaProvider", `
+{
+	"type": "object",
+	"additionalProperties": false,
+	"required": ["type", "alias"],
+	"properties": {
+		"type": { "type": "string", "enum": ["cloudflare", "recaptchav2"] },
+		"alias": { "type": "string" },
+		"site_key": { "type": "string" }
```
Comment (pkong-ds) [inline pkg/lib/config/captcha.go:26]: Actually am not very sure about this one: Is there a way to type force additionaled fields based on `type`?

Comment (louischan-oursky) [reply pkg/lib/config/captcha.go:26]: Yes, you use allOf, if, then. An example is pkg/lib/config/oauth.go:145

Comment (pkong-ds) [reply pkg/lib/config/captcha.go:26]: I see! https://json-schema.org/understanding-json-schema/reference/combining#allOf

Comment (pkong-ds) [reply pkg/lib/config/captcha.go:26]: But on golang level, how should `type force struct property based on other struct property` be done?

```
type Pet struct {
  Type PetType `json"type,omitempty"`
}

type PetType string

const (
  PetTypeDog PetType = "dog"
  PetTypeCat PetType = "cat"
)

type PetDog interface {
  Bark() nil
}

type PetCat interface {
  Meow() nil
}

// how should Bark() / Meow() method be required based on Pet.PetType?
// Actually, should dog and cat be their own structs, while pet be an interface?
// But if we implement dog and cat as their own structs, we would need *union* in another struct below? this makes things more complicated?

type Config struct {
  Pet (Dog | Cat) `json"pet"`
}
```

Comment (louischan-oursky) [reply pkg/lib/config/captcha.go:26]: You don't, and we don't. See pkg/lib/config/authentication_flow.go:893 The struct contains all fields, only some of them are used in a particular case.

**Comment by pkong-ds** [comment]
Missed secrets, lemme add them 🙏 

```
 [authgear.secrets.yaml](#authgearsecretsyaml)
      - [`type: cloudflare`](#type-cloudflare-1)
      - [`type: recaptchav2`](#type-recaptchav2-1)
```

File: pkg/lib/config/captcha.go [hunk 2]
```diff
@@ -5,24 +5,24 @@ var _ = Schema.Add("CaptchaConfig", `
...
 	"properties": {
-		"provider": { "$ref": "#/$defs/CaptchaProvider" }
+		"provider": { "$ref": "#/$defs/LegacyCaptchaProvider" }
 	}
 }
 `)
 
 type CaptchaConfig struct {
-	Provider *CaptchaProvider `json:"provider,omitempty"`
+	LegacyProvider *LegacyCaptchaProvider `json:"provider,omitempty"`
```
Comment (louischan-oursky) [inline pkg/lib/config/captcha.go:47]: - Rename the field LegacyProvider to Deprecated_Provider
- Rename the type LegacyCaptchaProvider to Deprecated_CaptchaProvider

File: pkg/lib/config/secret.go [hunk 3]
```diff
@@ -265,8 +265,8 @@ const (
 	// nolint: gosec
 	OAuthClientCredentialsKey SecretKey = "oauth.client_secrets"
 	// nolint: gosec
-	CaptchaCloudflareCredentialsKey  SecretKey = "captcha.cloudflare"
-	WhatsappOnPremisesCredentialsKey SecretKey = "whatsapp.on-premises"
+	LegacyCaptchaCloudflareCredentialsKey SecretKey = "captcha.cloudflare"
```
Comment (louischan-oursky) [inline pkg/lib/config/secret.go:271]: Rename LegacyCaptchaCloudflareCredentialsKey to Deprecated_CaptchaCloudflareCredentialsKey

File: pkg/lib/config/secret_captcha.go [hunk 4]
```diff
@@ -11,11 +11,11 @@ var _ = SecretConfigSchema.Add("CaptchaCloudflareCredentials", `
 }
 `)
 
-type CaptchaCloudflareCredentials struct {
+type LegacyCaptchaCloudflareCredentials struct {
```
Comment (louischan-oursky) [inline pkg/lib/config/secret_captcha.go:80]: Rename CaptchaCloudflareCredentials to Deprecated_CaptchaCloudflareCredentials

**Comment by pkong-ds** [comment]
@louischan-oursky updated, thanks!

**Review by louischan-oursky — APPROVED** [review]

File: pkg/lib/config/secret_captcha.go [hunk 5]
```diff
@@ -57,12 +61,11 @@ const (
...
 )
 
-func (c *CaptchaProvidersCredentials) SensitiveStrings() []string {
-	sensitiveStrings := make([]string, len(c.Items))
-	for i, cred := range c.Items {
-		sensitiveStrings[i] = cred.SecretKey
-	}
-	return sensitiveStrings
+func (c *CaptchaProvidersCredentials) SensitiveStrings() (sensitiveStrings []string) {
+	sensitiveStrings = slice.Map(c.Items, func(item CaptchaProvidersCredentialsItem) string {
```
Comment (louischan-oursky) [inline pkg/lib/config/secret_captcha.go:65]: The original pattern is OK. No need to rewrite.

**Pull request merged**